### ISSUE TITLE
Weighted SFT

### ISSUE THREAD

**Issue by RyanJFriedrich** [huggingface/trl#5761]
### Method description

I was in the process of writing a weighted training setup for SFT and thought it may add value to the TRL package. 

Normal SFT treats all assistant tokens equally. I personally found this frustrating because it makes mixed quality long form SFT less than ideal in a lot of ways. There is no real middle ground between a purely RL focused method or SFT focused method. The following implementation handles this at the assistant response level, but in the future I think adding the token level dynamics are perfectly feasible, and it's more an issue of data formatting. 

I'm proposing to add an OPTIONAL weight value to the JSON. It's a scalar -1 to 1 value. I expect the typical negative range not to go below -0.2 as negative values should be intentionally small. 

Something like the following :

{
  "messages": [
    {"role": "user", "content": "..."},
    {"role": "assistant", "content": "...", "weight": 1.0},
    {"role": "user", "content": "..."},
    {"role": "assistant", "content": "...", "weight": -0.1} 
  ]
}

The weight acts as a loss scalar. Positive values reinforce the response proportionally, while negative values invert the gradient direction. The above example of -0.1 acts to gently push the model away from that completion. 

It should be entirely backward compatible as this codepath only activates if a weight field is present in the dataset. I can separate the codepath such that the normal loss function logic is entirely untouched. Whatever is preferred. 

In terms of memory and compute versus standard SFT, there's pretty negligible overhead. 

A 20 turn roleplay conversation where turns 1 through 18 are mediocre but turns 19 and 20 contain the behavior you actually want to reinforce can now be used without contaminating your gradient with the mediocre turns.

Testing plan :

Unit test: verify `loss_weights` tensor is correctly produced by the collator from message-level `weight` fields
Unit test: verify weighted loss scales proportionally with weight magnitude
Integration test: train a small model with negative weights on a target token and verify suppression

Reference :
Reward-weighted regression applied to per-turn granularity in multi-turn conversations. Structurally parallels GRPO/RWR in data flow while using weighted cross-entropy instead of policy gradient.

This effectively extends the concept of OpenAI's fine tuning API from binary 0/1 weight logic per assistant message in to a continuous value. 

### Open source status

- [ ] The method implementation is available
- [ ] The model weights are available
- [ ] The training datasets are available

### Provide useful links for the implementation

I plan to write the implementation (~50 lines), mostly reusing existing SFT loss logic with a per-turn scalar applied. No external paper or reference implementation exists.  The closest conceptual precedent is reward-weighted regression applied at per turn granularity. Happy to open a PR directly.

**Comment by RyanJFriedrich** [comment]
I completed the implementation and tested it with a simple brainwashing experiment on Qwen2.5-0.5B-Instruct.

Phase 1: Train the model to express food preferences across 10 foods using randomized prompts and completions, all at weight 1.0. This establishes a roughly equal preference distribution.

Phase 2: Continue training. Hotdog completions get weight -0.1 (roughly 10% of examples) and everything else stays at 1.0.

Measurement: Average log-probability of "I really enjoy [food]." completions across held-out prompts the model never saw during training.

```
Food            Phase 1    Phase 2      Delta
------------------------------------------------
pizza           -0.6510    -0.6406    +0.0104
sushi           -0.5853    -0.7409    -0.1556
tacos           -0.7513    -0.7070    +0.0443
pasta           -0.6510    -0.5352    +0.1159
burgers         -0.8711    -0.7669    +0.1042
hotdogs         -0.5378    -8.4167    -7.8789 <-- HOTDOG
sandwiches      -0.6745    -0.6940    -0.0195
noodles         -0.6380    -0.6758    -0.0378
dumplings       -0.5951    -0.5977    -0.0026
steak           -0.6432    -0.6888    -0.0456
```

The negative signal was ~10% of phase 2 data at only -0.1 weight, yet produced a ~7.9 log-prob delta versus noise-level movement in other foods. Other preferences remained stable, meaning the negative weighting is surgical rather than destructive.

**Comment by matdou** [comment]
Hey, "no external paper" undersells it a bit, this is pretty well covered in the literature.

[ASFT (Zhu et al., 2025, ICLR 2026)](https://arxiv.org/abs/2509.23753) formalizes this kind of reward-weighted SFT within an RWR framework. And [Forgetting SFT (Ghahrizjani et al., 2025)](https://arxiv.org/abs/2508.04329) does the negative token thing at the token level... same idea, though weights there are computed automatically rather than specified by hand.

**Comment by RyanJFriedrich** [comment]
Thank you for bringing both papers to my awareness,

Neither of those papers covers a simple implementation. ASFT is technically offline. This requires frozen weights which which I prototyped early on and abandoned because it's a headache in practice. Forgetting SFT is pretty cool but requires training a model twice to calculate the gradient weights. It's better if you need automated token-level attribution. Most practitioners just want to say "this turn is good, this turn is mediocre, this turn is bad" without training an intermediate model.
 
I'm just offering a very small change to the JSONL format that allows every dataset on HF to contain the weight value.

Anyone can publish these as simple SFT datasets and the weight contributions are built in without calculating per token or needing to calculate exact gradient weights via the "Forgetting SFT" method.  

The code is complete and I've already used it to train a Gemma 31b model. It's a matter of writing the tests which I am admittedly being kind of lazy on. 

Oh and one last part, I don't know how to write the multi-GPU code path, so right now the code raises an exception due to lack of implementation.

**Comment by qgallouedec** [comment]
For ref, demand is still thin, waiting for signal, see  https://github.com/huggingface/trl/pull/5952#issuecomment-4764013527

**Comment by qgallouedec** [comment]
It's been a while since I was originally proposed, and this still is a two-person ask with no other reactions or issues. 

Plus, per-message weighted loss overlaps a fast-moving, unsettled research area (RWR/ASFT, DFT, token-level reweighting) where the automatic-weight variants dominate over hand-set scalars, while a `weight` dataset column is a public contract we'd be stuck supporting.
I don't think the maintenance cost balances the demand, consequently I'm going to close this issue and the associated PR, but happy to re-visit in the future if this post-training trick becomes more popular.

**Comment by RyanJFriedrich** [comment]
> It's been a while since I was originally proposed, and this still is a two-person ask with no other reactions or issues.
> 
> Plus, per-message weighted loss overlaps a fast-moving, unsettled research area (RWR/ASFT, DFT, token-level reweighting) where the automatic-weight variants dominate over hand-set scalars, while a `weight` dataset column is a public contract we'd be stuck supporting. I don't think the maintenance cost balances the demand, consequently I'm going to close this issue and the associated PR, but happy to re-visit in the future if this post-training trick becomes more popular.


Thank you for considering it. I'll maintain it for my own use for now. 

I attempted to maintain the "weight" contract as minimally as possible. The method basically only uses weight if it exists. The collation handles it being there, or not being there. It seemed not to care. The only complication would be adding "weight" later and it being a different variable.

Appreciate your time.
